src/plugins/kibana_utils/public/state_containers/create_state_container.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Subject } from 'rxjs';
+import { RecursiveReadonly } from '@kbn/utility-types';
+import {
+  PureTransitionsToTransitions,
+  PureTransition,
+  ReduxLikeStateContainer,
+  PureSelectorsToSelectors,
+} from './types';
+
+const $$observable = (typeof Symbol === 'function' && (Symbol as any).observable) || '@@observable';
+
+const freeze: <T>(value: T) => RecursiveReadonly<T> =
+  process.env.NODE_ENV !== 'production'
+    ? <T>(value: T): RecursiveReadonly<T> => {
+        if (!value) return value as RecursiveReadonly<T>;
+        if (typeof value === 'object') return Object.freeze({ ...value }) as RecursiveReadonly<T>;
+        else return value as RecursiveReadonly<T>;
+      }
+    : <T>(value: T) => value as RecursiveReadonly<T>;
+
+export const createStateContainer = <
+  State,
+  PureTransitions extends object,
+  PureSelectors extends object = {}
+>(
+  defaultState: State,
+  pureTransitions: PureTransitions,
+  pureSelectors: PureSelectors = {} as PureSelectors
+): ReduxLikeStateContainer<State, PureTransitions, PureSelectors> => {
+  const state$ = new Subject<RecursiveReadonly<State>>();
```

Comment:
it looks like it would be easier to have this `state$`a `BehaviorSubject` instead of `Subject`. 
Implementation will become a bit clearer as there would be no need to have separate `.state` value and no need to manually sync it with next `state$` values).

`get` will just become:
`get: () => state$.getValue();`
`state` could be defined as getter using object descriptors. 

Or maybe I am missing something else?

Reply:
On `.subscribe()` `BehaviourSubject` would immediately emit its last value? This is not what `state$` is expected to do and the Redux-style `.subsribe()` method.

Reply:
Refactored to use `BehaviourSubject` with `skip(1)`.